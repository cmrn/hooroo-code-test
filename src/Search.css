.Search-title {
  font-style: italic;
  font-weight: normal;
  font-size: 20px;
}

.Search-title > strong {
  font-style: initial;
  font-weight: bold;
}

.Search-sort {
  font-weight: bold;
}
