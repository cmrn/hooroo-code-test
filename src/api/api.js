import data from './data.json';

const SORT_HIGH = 'sort_high';
const SORT_LOW = 'sort_low';


function searchHotels(sortOrder) {
  const lowSortFunc = (a, b) => a.price > b.price;
  // return a promise to mimic an external API call
  return new Promise(resolve => {
    if(sortOrder === SORT_LOW) {
      resolve(data.sort(lowSortFunc));
    } else if(sortOrder === SORT_HIGH) {
      resolve(data.sort(lowSortFunc).reverse());
    } else {
      resolve(data);
    }
  });
}

export { searchHotels, SORT_HIGH, SORT_LOW };