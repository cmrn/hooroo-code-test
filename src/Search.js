import React from 'react';
import './Search.css';
import { searchHotels, SORT_LOW, SORT_HIGH } from './api';
import SearchResults from './SearchResults';

class Search extends React.Component {
  constructor(props) {
    super(props);
    this.state = { sort: SORT_HIGH, hotels: searchHotels(SORT_HIGH)};

    this.handleSortChange = this.handleSortChange.bind(this);
  }

  handleSortChange(event) {
    const newSort = event.target.value;
    this.setState({ sort: newSort, hotels: searchHotels(newSort) });
  }

  render() {
    return (
      <div className="Search">
        <div className="Search-header">
          <h2 className="Search-title">
            <strong>5</strong> hotels in <strong>Sydney</strong>.
          </h2>
          <div className="Search-sort">
            Sort by <select value={this.state.sort} onChange={this.handleSortChange}>
              <option value={SORT_HIGH}>Price high-low</option>
              <option value={SORT_LOW}>Price low-high</option>
            </select>
          </div>
        </div>
        <SearchResults results={this.state.hotels} />
      </div>
    );
  }
}

export default Search;