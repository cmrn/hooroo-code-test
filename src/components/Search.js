import React from 'react';
import PropTypes from 'prop-types';
import cnames from 'classnames';
import './Search.css';
import { SORT_LOW, SORT_HIGH } from '../api';
import SearchResults from './SearchResults';

const Search = ({ sort, hotels, onChange, className, ...props }) => { 
  return (
    <div className={cnames(className, "Search")} {...props}>
      <div className="Search-header">
        <h2 className="Search-title">
          <strong>{hotels.length}</strong> hotels in <strong>Sydney</strong>.
        </h2>
        <div className="Search-sort">
          Sort by <select value={sort} onChange={e => onChange(e.target.value)}>
            <option value={SORT_HIGH}>Price high-low</option>
            <option value={SORT_LOW}>Price low-high</option>
          </select>
        </div>
      </div>
      <SearchResults hotels={hotels} />
    </div>
  );
}

Search.propTypes = {
  sort: PropTypes.oneOf([SORT_LOW, SORT_HIGH]).isRequired,
  hotels: PropTypes.array.isRequired,
  onChange: PropTypes.func.isRequired,
  className: PropTypes.string,
}

export default Search;