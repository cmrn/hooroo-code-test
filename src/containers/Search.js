import React from 'react';
import { searchHotels, SORT_HIGH } from '../api';
import Search from '../components/Search';

class SearchContainer extends React.Component {
  constructor(props) {
    super(props);
    this.state = { sort: SORT_HIGH, hotels: [] };

    this.handleChange = this.handleChange.bind(this);
  }

  componentDidMount() {
    this.loadHotels(this.state.sort);
  }

  handleChange(newSort) {
    this.setState({ sort: newSort });
    this.loadHotels(newSort);
  }

  async loadHotels(sort) {
    const hotels = await searchHotels(sort);
    this.setState({ hotels });
  }

  render() {
    return <Search {...this.state} onChange={this.handleChange} />;
  }
}

export default SearchContainer;