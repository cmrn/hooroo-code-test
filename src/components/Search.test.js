import React from 'react';
import renderer from 'react-test-renderer';
import Search from './Search';
import { SORT_LOW } from '../api';

beforeEach(() => {
  Math.random = jest.fn(() => 0.0487723387081509); // react-star uses a random number
})

const hotels = [
  {
  "id": "cxd650nuyo",
  "title": "Courtyard by Marriott Sydney-North Ryde",
  "address": "7-11 Talavera Rd, North Ryde",
  "image": "https://unsplash.it/145/125/?random1",
  "rating": 4,
  "ratingType": "self",
  "promotion": "Exclusive Deal",
  "roomName": "Deluxe Balcony Room",
  "price": 329,
  "savings": 30,
  "freeCancellation": true
  }, {
  "id": "mesq6mggyn",
  "title": "Primus Hotel Sydney",
  "address": "339 Pitt St, Sydney",
  "image": "https://unsplash.it/145/125/?random2",
  "rating": 5,
  "ratingType": "self",
  "promotion": "Exclusive Deal",
  "roomName": "Deluxe King",
  "price": 375,
  "savings": 28,
  "freeCancellation": true
  }
];

it('renders no results when hotels is empty', () => {
  const tree = renderer.create(
    <Search sort={SORT_LOW} hotels={[]} onChange={()=>{}} />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

it('allows a className to be set', () => {
  const tree = renderer.create(
    <Search sort={SORT_LOW} hotels={[]} onChange={()=>{}} className="AdditionalClass" />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

it('passes other props through', () => {
  const tree = renderer.create(
    <Search sort={SORT_LOW} hotels={[]} onChange={()=>{}} aria-label="I should be on the div" />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders results when hotels is not empty', () => {
  const tree = renderer.create(
    <Search sort={SORT_LOW} hotels={hotels} onChange={()=>{}} />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});