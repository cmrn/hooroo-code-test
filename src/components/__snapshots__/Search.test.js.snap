// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`allows a className to be set 1`] = `
<div
  className="AdditionalClass Search"
>
  <div
    className="Search-header"
  >
    <h2
      className="Search-title"
    >
      <strong>
        0
      </strong>
       hotels in 
      <strong>
        Sydney
      </strong>
      .
    </h2>
    <div
      className="Search-sort"
    >
      Sort by 
      <select
        onChange={[Function]}
        value="sort_low"
      >
        <option
          value="sort_high"
        >
          Price high-low
        </option>
        <option
          value="sort_low"
        >
          Price low-high
        </option>
      </select>
    </div>
  </div>
  <div />
</div>
`;

exports[`passes other props through 1`] = `
<div
  aria-label="I should be on the div"
  className="Search"
>
  <div
    className="Search-header"
  >
    <h2
      className="Search-title"
    >
      <strong>
        0
      </strong>
       hotels in 
      <strong>
        Sydney
      </strong>
      .
    </h2>
    <div
      className="Search-sort"
    >
      Sort by 
      <select
        onChange={[Function]}
        value="sort_low"
      >
        <option
          value="sort_high"
        >
          Price high-low
        </option>
        <option
          value="sort_low"
        >
          Price low-high
        </option>
      </select>
    </div>
  </div>
  <div />
</div>
`;

exports[`renders no results when hotels is empty 1`] = `
<div
  className="Search"
>
  <div
    className="Search-header"
  >
    <h2
      className="Search-title"
    >
      <strong>
        0
      </strong>
       hotels in 
      <strong>
        Sydney
      </strong>
      .
    </h2>
    <div
      className="Search-sort"
    >
      Sort by 
      <select
        onChange={[Function]}
        value="sort_low"
      >
        <option
          value="sort_high"
        >
          Price high-low
        </option>
        <option
          value="sort_low"
        >
          Price low-high
        </option>
      </select>
    </div>
  </div>
  <div />
</div>
`;

exports[`renders results when hotels is not empty 1`] = `
<div
  className="Search"
>
  <div
    className="Search-header"
  >
    <h2
      className="Search-title"
    >
      <strong>
        2
      </strong>
       hotels in 
      <strong>
        Sydney
      </strong>
      .
    </h2>
    <div
      className="Search-sort"
    >
      Sort by 
      <select
        onChange={[Function]}
        value="sort_low"
      >
        <option
          value="sort_high"
        >
          Price high-low
        </option>
        <option
          value="sort_low"
        >
          Price low-high
        </option>
      </select>
    </div>
  </div>
  <div>
    <div
      className="Hotel"
    >
      <div
        className="Hotel-image"
      >
        <span
          className="Hotel-promotion"
        >
          Exclusive Deal
        </span>
        <img
          alt=""
          src="https://unsplash.it/145/125/?random1"
        />
      </div>
      <div
        className="Hotel-details"
      >
        <div
          className="Hotel-title"
        >
          <h3>
            Courtyard by Marriott Sydney-North Ryde
          </h3>
          <div
            aria-label="4 stars (self rated)"
            className="Hotel-rating Rating"
          >
            <div
              className={undefined}
              style={
                Object {
                  "overflow": "hidden",
                  "position": "relative",
                }
              }
            >
              <style
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "
                      .react-stars-00487723387081509:before {
                        position: absolute;
                        overflow: hidden;
                        display: block;
                        z-index: 1;
                        top: 0; left: 0;
                        width: 50%;
                        content: attr(data-forhalf);
                        color: #ffd700;
                    }",
                  }
                }
              />
              <span
                className=""
                data-forhalf="●"
                data-index={0}
                onClick={[Function]}
                onMouseLeave={[Function]}
                onMouseMove={[Function]}
                onMouseOver={[Function]}
                style={
                  Object {
                    "color": "#ffd700",
                    "cursor": "default",
                    "display": "block",
                    "float": "left",
                    "fontSize": "30px",
                    "overflow": "hidden",
                    "position": "relative",
                  }
                }
              >
                ●
              </span>
              <span
                className=""
                data-forhalf="●"
                data-index={1}
                onClick={[Function]}
                onMouseLeave={[Function]}
                onMouseMove={[Function]}
                onMouseOver={[Function]}
                style={
                  Object {
                    "color": "#ffd700",
                    "cursor": "default",
                    "display": "block",
                    "float": "left",
                    "fontSize": "30px",
                    "overflow": "hidden",
                    "position": "relative",
                  }
                }
              >
                ●
              </span>
              <span
                className=""
                data-forhalf="●"
                data-index={2}
                onClick={[Function]}
                onMouseLeave={[Function]}
                onMouseMove={[Function]}
                onMouseOver={[Function]}
                style={
                  Object {
                    "color": "#ffd700",
                    "cursor": "default",
                    "display": "block",
                    "float": "left",
                    "fontSize": "30px",
                    "overflow": "hidden",
                    "position": "relative",
                  }
                }
              >
                ●
              </span>
              <span
                className=""
                data-forhalf="●"
                data-index={3}
                onClick={[Function]}
                onMouseLeave={[Function]}
                onMouseMove={[Function]}
                onMouseOver={[Function]}
                style={
                  Object {
                    "color": "#ffd700",
                    "cursor": "default",
                    "display": "block",
                    "float": "left",
                    "fontSize": "30px",
                    "overflow": "hidden",
                    "position": "relative",
                  }
                }
              >
                ●
              </span>
              <span
                className=""
                data-forhalf="●"
                data-index={4}
                onClick={[Function]}
                onMouseLeave={[Function]}
                onMouseMove={[Function]}
                onMouseOver={[Function]}
                style={
                  Object {
                    "color": "#ccc",
                    "cursor": "default",
                    "display": "block",
                    "float": "left",
                    "fontSize": "30px",
                    "overflow": "hidden",
                    "position": "relative",
                  }
                }
              >
                ●
              </span>
            </div>
          </div>
        </div>
        <div
          className="Hotel-address"
        >
          7-11 Talavera Rd, North Ryde
        </div>
        <div
          className="Hotel-room"
        >
          Deluxe Balcony Room
        </div>
      </div>
      <div
        className="Hotel-cancellation"
      >
        Free cancellation
      </div>
      <div
        className="Hotel-price"
      >
        <div
          className="Hotel-priceDescription"
        >
          1 night total (AUD)
        </div>
        <div>
          <span
            className="Hotel-currencySymbol"
          >
            $
          </span>
           
          <span
            className="Hotel-priceNumber"
          >
            329
          </span>
        </div>
        <div
          className="Hotel-savings"
        >
          Save $
          30
          ~
        </div>
      </div>
    </div>
    <div
      className="Hotel"
    >
      <div
        className="Hotel-image"
      >
        <span
          className="Hotel-promotion"
        >
          Exclusive Deal
        </span>
        <img
          alt=""
          src="https://unsplash.it/145/125/?random2"
        />
      </div>
      <div
        className="Hotel-details"
      >
        <div
          className="Hotel-title"
        >
          <h3>
            Primus Hotel Sydney
          </h3>
          <div
            aria-label="5 stars (self rated)"
            className="Hotel-rating Rating"
          >
            <div
              className={undefined}
              style={
                Object {
                  "overflow": "hidden",
                  "position": "relative",
                }
              }
            >
              <style
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "
                      .react-stars-00487723387081509:before {
                        position: absolute;
                        overflow: hidden;
                        display: block;
                        z-index: 1;
                        top: 0; left: 0;
                        width: 50%;
                        content: attr(data-forhalf);
                        color: #ffd700;
                    }",
                  }
                }
              />
              <span
                className=""
                data-forhalf="●"
                data-index={0}
                onClick={[Function]}
                onMouseLeave={[Function]}
                onMouseMove={[Function]}
                onMouseOver={[Function]}
                style={
                  Object {
                    "color": "#ffd700",
                    "cursor": "default",
                    "display": "block",
                    "float": "left",
                    "fontSize": "30px",
                    "overflow": "hidden",
                    "position": "relative",
                  }
                }
              >
                ●
              </span>
              <span
                className=""
                data-forhalf="●"
                data-index={1}
                onClick={[Function]}
                onMouseLeave={[Function]}
                onMouseMove={[Function]}
                onMouseOver={[Function]}
                style={
                  Object {
                    "color": "#ffd700",
                    "cursor": "default",
                    "display": "block",
                    "float": "left",
                    "fontSize": "30px",
                    "overflow": "hidden",
                    "position": "relative",
                  }
                }
              >
                ●
              </span>
              <span
                className=""
                data-forhalf="●"
                data-index={2}
                onClick={[Function]}
                onMouseLeave={[Function]}
                onMouseMove={[Function]}
                onMouseOver={[Function]}
                style={
                  Object {
                    "color": "#ffd700",
                    "cursor": "default",
                    "display": "block",
                    "float": "left",
                    "fontSize": "30px",
                    "overflow": "hidden",
                    "position": "relative",
                  }
                }
              >
                ●
              </span>
              <span
                className=""
                data-forhalf="●"
                data-index={3}
                onClick={[Function]}
                onMouseLeave={[Function]}
                onMouseMove={[Function]}
                onMouseOver={[Function]}
                style={
                  Object {
                    "color": "#ffd700",
                    "cursor": "default",
                    "display": "block",
                    "float": "left",
                    "fontSize": "30px",
                    "overflow": "hidden",
                    "position": "relative",
                  }
                }
              >
                ●
              </span>
              <span
                className=""
                data-forhalf="●"
                data-index={4}
                onClick={[Function]}
                onMouseLeave={[Function]}
                onMouseMove={[Function]}
                onMouseOver={[Function]}
                style={
                  Object {
                    "color": "#ffd700",
                    "cursor": "default",
                    "display": "block",
                    "float": "left",
                    "fontSize": "30px",
                    "overflow": "hidden",
                    "position": "relative",
                  }
                }
              >
                ●
              </span>
            </div>
          </div>
        </div>
        <div
          className="Hotel-address"
        >
          339 Pitt St, Sydney
        </div>
        <div
          className="Hotel-room"
        >
          Deluxe King
        </div>
      </div>
      <div
        className="Hotel-cancellation"
      >
        Free cancellation
      </div>
      <div
        className="Hotel-price"
      >
        <div
          className="Hotel-priceDescription"
        >
          1 night total (AUD)
        </div>
        <div>
          <span
            className="Hotel-currencySymbol"
          >
            $
          </span>
           
          <span
            className="Hotel-priceNumber"
          >
            375
          </span>
        </div>
        <div
          className="Hotel-savings"
        >
          Save $
          28
          ~
        </div>
      </div>
    </div>
  </div>
</div>
`;
